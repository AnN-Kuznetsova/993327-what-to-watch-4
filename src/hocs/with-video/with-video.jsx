import PropTypes from "prop-types";
import React, {PureComponent, createRef} from "react";


export const VideoPlayerMode = {
  PREVIEW: `preview`,
  FULL_SCREEN: `full-screen`,
};


export const videoOptions = {
  [VideoPlayerMode.PREVIEW]: {
    isAutoPlay: false,
    isSound: false,
    videoHeight: 175,
  },
  [VideoPlayerMode.FULL_SCREEN]: {},
};


export const withVideo = (Component) => {
  class WithVideo extends PureComponent {
    constructor(props) {
      super(props);

      this._videoRef = createRef();

      this.state = {
        progress: 0,
        isLoading: true,
      };
    }

    componentDidMount() {
      const {src, playerMode} = this.props;
      const {isSound} = videoOptions[playerMode];
      const video = this._videoRef.current;

      video.src = src;
      video.muted = !isSound;

      video.oncanplaythrough = () => this.setState({
        isLoading: false,
      });

      video.ontimeupdate = () => this.setState({
        progress: Math.floor(video.currentTime),
      });
    }

    componentDidUpdate() {
      const video = this._videoRef.current;
      const {playerMode} = this.props;

      if (this.props.isPlaying) {
        video.play();
      } else {
        if (playerMode === VideoPlayerMode.PREVIEW) {
          video.load();
        } else {
          video.pause();
        }
      }
    }

    componentWillUnmount() {
      const video = this._videoRef.current;

      video.oncanplaythrough = null;
      video.ontimeupdate = null;
      video.src = ``;
    }

    render() {
      const {posterUrl, playerMode, isPlaying} = this.props;
      const {videoHeight} = videoOptions[playerMode];

      return (
        <Component
          {...this.props}
          playerMode={playerMode}
        >
          <video
            ref={this._videoRef}
            className="player__video"
            poster={posterUrl}
            autoPlay={isPlaying}
            style={{height: videoHeight}}
          />
        </Component>
      );
    }
  }


  WithVideo.propTypes = {
    src: PropTypes.string.isRequired,
    posterUrl: PropTypes.string.isRequired,
    playerMode: PropTypes.string.isRequired,
    isPlaying: PropTypes.bool.isRequired,
  };


  return WithVideo;
};
