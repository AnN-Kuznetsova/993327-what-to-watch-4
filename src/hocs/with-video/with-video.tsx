
import * as React, {PureComponent, createRef} from "react";

import {MoviePropType} from "../../prop-types";
import {PageType, AppRoute} from "../../const";
import {history} from "../../history";


export const VideoPlayerMode = {
  PREVIEW: `PREVIEW`,
  SMALL_SCREEN: `SMALL_SCREEN`,
  FULL_SCREEN: `FULL_SCREEN`,
};


export const videoOptions = {
  [VideoPlayerMode.PREVIEW]: {
    isAutoPlay: false,
    isSound: false,
  },
  [VideoPlayerMode.SMALL_SCREEN]: {
    isAutoPlay: true,
    isSound: true,
  },
  [VideoPlayerMode.FULL_SCREEN]: {
    isAutoPlay: true,
    isSound: true,
  },
};

export const VideoPlayerStatus = {
  ON_AUTOPLAY: `on-autoplay`,
  ON_PLAY: `on-play`,
  ON_PAUSE: `on-pause`,
  ON_RESET: `on-reset`,
};


export const withVideo = (Component) => {
  class WithVideo extends PureComponent {
    constructor(props) {
      super(props);

      this._videoRef = createRef();
      this._duration = null;

      this.state = {
        progress: 0,
        isLoading: true,
      };
    }

    componentDidMount() {
      const {src, playerMode} = this.props;
      const {isSound} = videoOptions[playerMode];
      const video = this._videoRef.current;

      video.src = src;
      video.muted = !isSound;

      video.oncanplaythrough = () => {
        this.setState({
          isLoading: false,
        });

        this._duration = video.duration;
      };

      video.ontimeupdate = () => this.setState({
        progress: Math.floor(video.currentTime),
      });
    }

    componentDidUpdate() {
      const video = this._videoRef.current;
      const {playerMode} = this.props;
      const {isLoading} = this.state;

      if (!isLoading) {
        if (this.getPlayingValue()) {
          const playPromise = video.play();
          if (playPromise) {
            playPromise.catch(()=>{
              this.props.setVideoPlayerStatus(VideoPlayerStatus.ON_PAUSE);
            });
          }
        } else {
          if (playerMode === VideoPlayerMode.PREVIEW) {
            video.load();
          } else {
            video.pause();
          }
        }
      }
    }

    componentWillUnmount() {
      const video = this._videoRef.current;

      video.oncanplaythrough = null;
      video.ontimeupdate = null;
      video.src = ``;
    }

    getPlayingValue() {
      switch (this.props.playerStatus) {
        case VideoPlayerStatus.ON_AUTOPLAY:
          return videoOptions[this.props.playerMode].isAutoPlay;
        case VideoPlayerStatus.ON_PLAY:
          return true;
        case VideoPlayerStatus.ON_PAUSE:
        case VideoPlayerStatus.ON_RESET:
          return false;
        default:
          throw new Error(`Unknown type of playing state.`);
      }
    }

    handlePlayButtonClick() {
      if (this.getPlayingValue()) {
        this.props.setVideoPlayerStatus(VideoPlayerStatus.ON_PAUSE);
      } else {
        this.props.setVideoPlayerStatus(VideoPlayerStatus.ON_PLAY);
      }
    }

    handleExitButtonClick(event) {
      event.preventDefault();

      const {activePage, prevPage} = this.props;

      if (activePage === PageType.PLAYER) {
        const page = prevPage === PageType.MAIN ? AppRoute.MAIN : AppRoute.FILM.replace(`:id`, this.props.activeMovie.id);
        this.props.onChangePage(prevPage);
        history.push(page);
      } else {
        this.props.onChangePage(activePage);
      }
      this.props.setVideoPlayerStatus(VideoPlayerStatus.ON_AUTOPLAY);
    }

    handleFullScreenButtonClick(mode) {
      this.props.setVideoPlayerMode(mode);
    }

    render() {
      const {
        posterUrl,
        playerMode,
      } = this.props;
      const {
        isLoading,
        progress,
      } = this.state;
      const isPlaying = this.getPlayingValue();

      return (
        <Component
          {...this.props}
          playerMode={playerMode}
          isLoading={isLoading}
          isPlaying={isPlaying}
          duration={this._duration}
          progress={progress || 0}
          onPlayButtonClick={this.handlePlayButtonClick.bind(this)}
          onExitButtonClick={this.handleExitButtonClick.bind(this)}
          onFullScreenButtonClick={this.handleFullScreenButtonClick.bind(this)}
        >
          <video
            ref={this._videoRef}
            className="player__video"
            poster={posterUrl}
            autoPlay={isPlaying}
          />
        </Component>
      );
    }
  }


  WithVideo.propTypes = {
    src: PropTypes.string.isRequired,
    posterUrl: PropTypes.string.isRequired,
    playerMode: PropTypes.string.isRequired,
    activeMovie: MoviePropType.isRequired,
    activePage: PropTypes.string,
    prevPage: PropTypes.string.isRequired,
    onChangePage: PropTypes.func.isRequired,
    setVideoPlayerMode: PropTypes.func.isRequired,
    setVideoPlayerStatus: PropTypes.func.isRequired,
    playerStatus: PropTypes.string.isRequired,
  };


  return WithVideo;
};
