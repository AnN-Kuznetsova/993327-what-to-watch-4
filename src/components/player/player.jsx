import PropTypes from "prop-types";
import React from "react";
import {connect} from "react-redux";

import {ActionCreator} from "../../reducers/reducer";
import {withVideo, VideoPlayerMode} from "../../hocs/with-video/with-video";


const getFormatedTimeLeft = (time) => {
  const seconds = Math.round(time % 60);
  const minutes = Math.floor(time / 60);
  const hours = Math.floor(minutes / 60);
  return `${hours}:${minutes}:${seconds} `;
};


const PlayerComponent = (props) => {
  const {
    playerMode,
    isPlaying,
    duration,
    progress,
    children,
    onPlayButtonClick,
    onExitButtonClick,
    onFullScreenButtonClick,
  } = props;
  const timeLeft = duration ? getFormatedTimeLeft(duration - progress) : 0;
  const progressValue = duration ? progress * 100 / duration : 0;

  const getPlayer = () => {
    switch (playerMode) {
      case VideoPlayerMode.SMALL_SCREEN:
      case VideoPlayerMode.FULL_SCREEN:
        return (
          <div className={playerMode === VideoPlayerMode.FULL_SCREEN ? `player` : ``}>
            {children}

            <button
              type="button"
              className="player__exit"
              onClick={onExitButtonClick}
            >Exit</button>

            <div
              className="player__controls"
              style={{position: `${playerMode === VideoPlayerMode.FULL_SCREEN ? `absolute` : `relative`}`}}>
              <div className="player__controls-row">
                <div className="player__time">
                  <progress className="player__progress" value={progressValue} max="100" />
                  <div className="player__toggler" style={{left: `${progressValue}%`}}>Toggler</div>
                </div>
                <div className="player__time-value">{timeLeft}</div>
              </div>

              <div className="player__controls-row">
                <button
                  type="button"
                  className="player__play"
                  onClick={onPlayButtonClick}
                >
                  {isPlaying &&
                  <svg viewBox="0 0 14 21" width="14" height="21">
                    <use xlinkHref="#pause" />
                  </svg>
                  ||
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s" />
                  </svg>
                  }
                  <span>{isPlaying && `Pause` || `Play`}</span>
                </button>
                <div className="player__name">Transpotting</div>

                {playerMode === VideoPlayerMode.SMALL_SCREEN &&
                <button
                  type="button"
                  className="player__full-screen"
                  onClick={onFullScreenButtonClick}
                >
                  <svg viewBox="0 0 27 27" width="27" height="27">
                    <use xlinkHref="#full-screen" />
                  </svg>
                  <span>Full screen</span>
                </button>}
              </div>
            </div>
          </div>
        );

      case VideoPlayerMode.PREVIEW:
      default:
        return (
          <React.Fragment>
            {children}
          </React.Fragment>
        );
    }
  };

  return getPlayer();
};


PlayerComponent.propTypes = {
  playerMode: PropTypes.string.isRequired,
  isPlaying: PropTypes.bool.isRequired,
  duration: PropTypes.number,
  progress: PropTypes.number.isRequired,
  activePage: PropTypes.string,
  prevPage: PropTypes.string.isRequired,
  playerStartTime: PropTypes.number.isRequired,
  onChangePage: PropTypes.func.isRequired,
  setPlayerStartTime: PropTypes.func.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};


const mapStateToProps = (state) => ({
  activePage: state.activePage,
  prevPage: state.prevPage,
  playerStartTime: state.playerStartTime,
});

const mapDispatchToProps = (dispatch) => ({
  onChangePage(newPage) {
    dispatch(ActionCreator.changeActivePage(newPage));
  },
  setPlayerStartTime(currentTime) {
    dispatch(ActionCreator.setPlayerStartTime(currentTime));
  },
});

const PlayerComponentWithVideo = withVideo(PlayerComponent);
const PlayerWithVideo = connect(mapStateToProps, mapDispatchToProps)(PlayerComponentWithVideo);


export {
  PlayerComponent,
  PlayerComponentWithVideo,
  PlayerWithVideo,
};
