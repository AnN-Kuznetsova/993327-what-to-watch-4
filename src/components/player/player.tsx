import * as React from "react";
import {connect} from "react-redux";

import {ActionCreator} from "../../reducers/application/application";
import {MovieType} from "../../types";
import {getActivePage, getPrevPage, getActiveMovie} from "../../reducers/application/selectors";
import {withVideo, VideoPlayerMode} from "../../hocs/with-video/with-video";


interface Props {
  playerMode: string;
  isLoading: boolean;
  isPlaying: boolean;
  duration: number;
  progress: number;
  activeMovie: MovieType;
  activePage: string;
  prevPage: string;
  onChangePage: (newPage: string) => void;
  children: React.ReactNode | React.ReactNode[];
  onPlayButtonClick: () => void;
  onExitButtonClick: () => void;
  onFullScreenButtonClick: (mode: string) => void;
}


const getFormatedTimeLeft = (time) => {
  const seconds = Math.round(time % 60);
  const minutes = Math.floor(time / 60);
  const hours = Math.floor(minutes / 60);
  return `${hours}:${minutes}:${seconds} `;
};


const PlayerComponent: React.FunctionComponent<Props> = (props: Props) => {
  const {
    playerMode,
    isLoading,
    isPlaying,
    duration,
    progress,
    activeMovie,
    children,
    onPlayButtonClick,
    onExitButtonClick,
    onFullScreenButtonClick,
  } = props;
  const timeLeft = duration ? getFormatedTimeLeft(duration - progress) : 0;
  const progressValue = duration ? progress * 100 / duration : 0;
  const playerRef: React.RefObject<HTMLDivElement> = React.createRef();

  const handleFullScreenButtonClick = () => {
    if (playerMode === VideoPlayerMode.SMALL_SCREEN) {
      playerRef.current.requestFullscreen();
      onFullScreenButtonClick(VideoPlayerMode.FULL_SCREEN);
    } else if (playerMode === VideoPlayerMode.FULL_SCREEN) {
      document.exitFullscreen();
      onFullScreenButtonClick(VideoPlayerMode.SMALL_SCREEN);
    }
  };

  const getPlayer = () => {
    switch (playerMode) {
      case VideoPlayerMode.SMALL_SCREEN:
      case VideoPlayerMode.FULL_SCREEN:
        return (
          <div
            className="player"
            ref={playerRef}>
            {children}

            <button
              type="button"
              className="player__exit"
              onClick={onExitButtonClick}
            >Exit</button>

            <div className="player__controls" >
              <div className="player__controls-row">
                <div className="player__time">
                  <progress className="player__progress" value={progressValue} max="100" />
                  <div className="player__toggler" style={{left: `${progressValue}%`}}>Toggler</div>
                </div>
                <div className="player__time-value">{timeLeft}</div>
              </div>

              <div className="player__controls-row">
                <button
                  type="button"
                  className="player__play"
                  disabled={isLoading}
                  onClick={onPlayButtonClick}
                >
                  {isPlaying &&
                  <svg viewBox="0 0 14 21" width="14" height="21">
                    <use xlinkHref="#pause" />
                  </svg>
                  ||
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s" />
                  </svg>
                  }
                  <span>{isPlaying && `Pause` || `Play`}</span>
                </button>
                <div className="player__name">{activeMovie.title}</div>

                <button
                  className="player__full-screen"
                  onClick={handleFullScreenButtonClick}
                >
                  <svg viewBox="0 0 27 27" width="27" height="27">
                    <use xlinkHref="#full-screen" />
                  </svg>
                  <span>Full screen</span>
                </button>
              </div>
            </div>
          </div>
        );

      case VideoPlayerMode.PREVIEW:
      default:
        return (
          <React.Fragment>
            {children}
          </React.Fragment>
        );
    }
  };

  return getPlayer();
};


const mapStateToProps = (state) => ({
  activePage: getActivePage(state),
  activeMovie: getActiveMovie(state),
  prevPage: getPrevPage(state),
});

const mapDispatchToProps = (dispatch) => ({
  onChangePage(newPage) {
    dispatch(ActionCreator.changeActivePage(newPage));
  },
});

const PlayerComponentWithVideo = withVideo(PlayerComponent);
const PlayerWithVideo = connect(mapStateToProps, mapDispatchToProps)(PlayerComponentWithVideo);


export {
  PlayerComponent,
  PlayerComponentWithVideo,
  PlayerWithVideo,
};
