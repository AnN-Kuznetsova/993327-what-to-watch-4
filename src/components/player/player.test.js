import React from "react";
import renderer from "react-test-renderer";

import {Player} from "./player";
import {VideoPlayerMode} from "../../hocs/with-video/with-video";


const nodeMock = {
  createNodeMock: () => {
    return {};
  }
};

const props = {
  playerMode: VideoPlayerMode.PREVIEW,
};


describe(`Render Player`, () => {
  it(`Should match with snapshot when player mode is "PREVIEW"`, () => {
    const videoPlayerSnapshot = renderer.create(
        <Player {...props} >
          <video />
        </Player>, nodeMock
    ).toJSON();

    expect(videoPlayerSnapshot).toMatchSnapshot();
  });
});
