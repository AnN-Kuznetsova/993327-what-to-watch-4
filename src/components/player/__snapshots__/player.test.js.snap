// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Player Should match with snapshot when player mode is "PREVIEW" 1`] = `
<div
  className=""
>
  <video />
</div>
`;
